### ISSUE TITLE
poor ichunked performance

### ISSUE THREAD

**Issue by seanmacavaney** [more-itertools/more-itertools#576]
It looks like the implementation of `ichunked` is considerably less efficient than `chunked`:

```python
>>> from tqdm import tqdm
>>> from more_itertools import chunked, ichunked
>>> for _ in ichunked(tqdm(range(3_000_000)), 1000):
...   pass
 23%|  698461/3000000 [00:05<00:36, 63231.90it/s]
 32%|  948724/3000000 [00:10<00:48, 42513.00it/s]
 38%| 1137833/3000000 [00:15<00:55, 33747.15it/s]
 43%| 1287378/3000000 [00:20<00:59, 28855.96it/s]
 48%| 1425604/3000000 [00:25<01:01, 25466.22it/s]
 51%| 1539932/3000000 [00:30<01:02, 23219.40it/s]
 55%| 1652571/3000000 [00:35<01:05, 20458.89it/s]
 58%| 1751094/3000000 [00:40<01:02, 19879.07it/s]
 62%| 1847406/3000000 [00:45<01:01, 18642.17it/s]
 65%| 1938015/3000000 [00:50<01:00, 17611.84it/s]
 68%| 2040571/3000000 [00:55<00:57, 16569.79it/s]
 70%| 2113594/3000000 [01:00<00:55, 15867.30it/s]
# ^ getting progressively slower over time

>>> for _ in chunked(tqdm(range(3_000_000)), 1000):
...   pass
100%| 3000000/3000000 [00:00<00:00, 6905401.43it/s]
# ^ finishes in under 1sec
```

The above behaviour is also present when:
 - Not using `tqdm`
 - Fully consuming the iterable in the loop

I'm not sure how practical it is to address this, but it's probably worthy of documentation. It had a noticeable impact on our tool as part of a larger pipeline (see https://github.com/terrier-org/pyterrier/pull/238).

Seen using:
 - `more-itertools==8.8.0`
 - `python==3.7`

**Comment by bbayles** [comment]
A PR is welcome for:
(a) Documenting this
(b) Making it better

Thanks for the report!

**Comment by bbayles** [comment]
@seanmacavaney, how important is out-of-order reading of chunks to you?

I am considering changing the function to require reading in order. Storing the items seems to be the cause of the performance degradation - the repeated calls to `tee` create more and more `deque`-like objects.

I'll see how the function is used on GitHub. If the usage changes then the change will be made in a new major version.

**Comment by seanmacavaney** [comment]
In our application, out-of-order reading isn't important. We just went with `ichunked` because we figured it'd be more efficient since it doesn't need to create an extra list.

I think it's reasonable to require the chunks to be read in order. It doesn't seem like there'd be a way to allow that without storing results internally somehow -- in which case you might as well just use `chunked`.

**Comment by bbayles** [comment]
Addressed in #587.
